src/arith_uint256.cpp

```diff
@@ -176,7 +176,7 @@ unsigned int base_uint<BITS>::bits() const
     for (int pos = WIDTH - 1; pos >= 0; pos--) {
         if (pn[pos]) {
             for (int nbits = 31; nbits > 0; nbits--) {
-                if (pn[pos] & 1 << nbits)
+                if (pn[pos] & 1U << nbits)
```

Comment:
`pn[...]` is uint32_t, how is this ever undefined behavior?
I don't think C ever "promotes" uint32_t to int.

Reply:
@laanwj With parenthesis it's `pn[pos & (1 << nbits)`; I believe the type of `(1 << nbits)` is indeed `int`, which in case of `nbits == 31` is exceeding the size of the type.

Reply:
Can you add a test that fails without this fix?

Reply:
no, he can't—the whole problem, the whole shitshow with C and C++ is that you can't predict what will happen with undefined behavior—it can just work with one compiler or version and commit war crimes the next one

Reply:
@promag How would one test for the presence of undefined behaviour? :-)

If you are referring to the sanitisers there are no sanitisers flagging this UB AFAIK.